### PR TITLE
sys/net/fib: fix compilation with -Wcast-align

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

As the title says.

### Testing procedure

It might make sense to verify that a `memcpy()` is indeed needed. If one could deduce that alignment is provided at the caller site under all circumstances, the `memcpy()` introduced unneeded overhead.

### Issues/PRs references

Split out of https://github.com/RIOT-OS/RIOT/pull/14955

### CODE REVIEW

File: sys/net/network_layer/fib/fib.c [hunk 1]
```diff
@@ -363,10 +363,12 @@ static int fib_signal_rp(fib_table_t *table, uint16_t type, uint8_t *dat,
                 }
             }
             else {
-                fib_sr_t *temp_sr = (fib_sr_t *)dat;
-                size_t dat_size_in_bits = temp_sr->sr_dest->address->address_size << 3;
+                fib_sr_t temp_sr;
+                assert(dat_size >= sizeof(temp_sr));
+                memcpy(&temp_sr, dat, sizeof(temp_sr));
```
Comment (benpicco) [inline sys/net/network_layer/fib/fib.c:368]: Why copy the whole `fib_sr_t` if we only are interested in `sr_dest`?

```C
fib_sr_entry_t *sr_dest;
memcpy(sr_dest, &temp_sr->sr_dest, sizeof(sr_dest));
```

Comment (maribu) [reply sys/net/network_layer/fib/fib.c]: Like this?

File: sys/net/network_layer/fib/fib.c [hunk 2]
```diff
@@ -363,10 +363,12 @@ static int fib_signal_rp(fib_table_t *table, uint16_t type, uint8_t *dat,
                 }
             }
             else {
-                fib_sr_t *temp_sr = (fib_sr_t *)dat;
-                size_t dat_size_in_bits = temp_sr->sr_dest->address->address_size << 3;
+                fib_sr_entry_t sr_dest;
+                assert(dat_size >= sizeof(fib_sr_t));
+                memcpy(&sr_dest, ((fib_sr_t *)(uintptr_t)dat)->sr_dest, sizeof(fib_entry_t));
+                size_t dat_size_in_bits = sr_dest.address->address_size << 3;
```
Comment (benpicco) [inline sys/net/network_layer/fib/fib.c:369]: I mean like 
```suggestion
                fib_sr_t *temp_sr = (fib_sr_t *)(uintptr_t)dat;
                fib_sr_entry_t *sr_dest;
                memcpy(&sr_dest, &temp_sr->sr_dest, sizeof(sr_dest));
                size_t dat_size_in_bits = sr_dest->address->address_size << 3;
```

Comment (maribu) [reply sys/net/network_layer/fib/fib.c:369]: But I guess if we trust the byte array to contain valid pointers within the local memory, we can just as well trust it to be correctly aligned.

**Comment by maribu** [comment]
```
$ BOARD=nucleo-f767zi make -C tests/unittests tests-fib DEVELHELP=1 flash test
[...]
Help: Press s to start test, r to print it is ready
READY
s
START
....................
OK (20 tests)
```

and

```
$ BOARD=nucleo-f767zi make -C tests/unittests tests-fib_sr DEVELHELP=1 flash test
[...]
READY
s
START
............
OK (12 tests)
```

So I guess there is no alignment issue, as the `assert()` didn't blow.

**Review by benpicco — APPROVED** [review]

File: sys/net/network_layer/fib/fib.c [hunk 3]
```diff
@@ -956,7 +967,8 @@ int fib_sr_entry_append(fib_table_t *table, fib_sr_t *fib_sr,
     fib_sr_entry_t *elt = NULL;
     LL_FOREACH(fib_sr->sr_path, elt) {
         size_t addr_size_match = addr_size << 3;
-        if (universal_address_compare(elt->address, addr, &addr_size_match) == UNIVERSAL_ADDRESS_EQUAL) {
+        if (UNIVERSAL_ADDRESS_EQUAL == universal_address_compare(elt->address, addr,
+                                                                 &addr_size_match)) {
```
Comment (benpicco) [inline sys/net/network_layer/fib/fib.c:971]: I'm not complaining about the change, but CI is warning about *that*?
Yoda conditions can be helpful, but do we really need CI to enforce them? Assignments inside `if` conditions already raise a warning, so this seems unnecessary.

**Comment by maribu** [comment]
I guess it is time for my to package `vera++` for Alpine so that I can locally confirm that the CI is happy, rather than three-times force-pushing things that I overlooked. But now the CI should like this PR.

File: sys/net/network_layer/fib/fib.c [hunk 4]
```diff
@@ -144,9 +148,11 @@ static int fib_find_entry(fib_table_t *table, uint8_t *dst, size_t dst_size,
                 /* we try to find the most fitting prefix */
                 if (ret_comp == UNIVERSAL_ADDRESS_MATCHING_PREFIX) {
                     if (table->data.entries[i].global_flags & FIB_FLAG_NET_PREFIX_MASK) {
-                        /* we shift the most upper flag byte back to get the number of prefix bits */
-                        size_t global_prefix_len = (table->data.entries[i].global_flags
-                                                    & FIB_FLAG_NET_PREFIX_MASK) >> FIB_FLAG_NET_PREFIX_SHIFT;
+                        /* we shift the most upper flag byte back to get the number of prefix
+                         * bits */
+                        size_t global_prefix_len = table->data.entries[i].global_flags;
```
Comment (benpicco) [inline sys/net/network_layer/fib/fib.c:153]:
```suggestion
                        uint32_t global_prefix_len = table->data.entries[i].global_flags;
```

Comment (maribu) [reply sys/net/network_layer/fib/fib.c:153]: good catch!

**Comment by maribu** [comment]
rabased and squashed

**Comment by maribu** [comment]
Thanks for the review!

**Pull request merged**